astra/src/main/java/com/slack/astra/zipkinApi/ZipkinService.java

```diff
@@ -242,6 +281,93 @@ public HttpResponse getTraceByTraceId(
     List<LogWireMessage> messages = searchResultToLogWireMessage(searchResult);
     String output = convertLogWireMessageToZipkinSpan(messages);
 
+    if (userRequest.isPresent() && userRequest.get() && !output.isEmpty()) {
+      // Save the trace data to S3 if the custom header is present
+      saveDataToS3(traceId, output);
+    }
     return HttpResponse.of(HttpStatus.OK, MediaType.JSON_UTF_8, output);
   }
+
+  @VisibleForTesting
+  protected static byte[] compressJsonData(String jsonData) throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    try (GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream)) {
+      gzipOutputStream.write(jsonData.getBytes(StandardCharsets.UTF_8));
+    }
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  @VisibleForTesting
+  protected static String decompressJsonData(byte[] compressedData) {
+    assert compressedData != null && compressedData.length > 0;
+
+    try (GZIPInputStream gzipInputStream =
+        new GZIPInputStream(new ByteArrayInputStream(compressedData))) {
+      return new String(gzipInputStream.readAllBytes(), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      LOG.error("Error decompressing JSON data", e);
+      throw new RuntimeException("Failed to decompress JSON data", e);
+    }
+  }
+
+  protected String retrieveDataFromS3(String traceId, long dataFreshnessInMinutes) {
+    assert traceId != null && !traceId.isEmpty();
+
+    try {
+      // If data is fresh skip S3 retrieval
+      long currentTime = Instant.now().toEpochMilli();
+
+      HeadObjectResponse response =
+          blobStore.getFileMetadata(
+              String.format("%s/%s/traceData.json.gz", TRACE_CACHE_PREFIX, traceId));
+      long lastModified = response.lastModified().toEpochMilli();
+      if (currentTime - lastModified < dataFreshnessInMinutes * 60 * 1000) {
+        return null; // Data is still getting updated, check on cache and live nodes
+      }
+
+      // Retrieve the compressed trace data from S3
+      java.nio.file.Path tempDir = Files.createTempDirectory("");
+      blobStore.download(String.format("%s/%s", TRACE_CACHE_PREFIX, traceId), tempDir);
```

Comment:
It feels a little heavy handed to download the file to a directory in order to be able to send it back out, given we know the exact key.

Maybe we could do something like 
```
s3AsyncClient.getObject(GetObjectRequest.builder().bucket(bucketName).key(prefix+"/"+"...").build(), AsyncResponseTransformer.toBlockingInputStream())
```
to get an inputstream of it and wrap that with a gzip decompression?

Reply:
+1. This could be more efficient and also saves disk io. One more option to explore is since the data output format is the same if we can redirect with a signed S3 URL.